File: Interview/UIView CALayer/Demo素材/2源代码/15-小画板(线宽)/小画板/ViewController.m
//
//  ViewController.m
//  小画板
//
//  Created by Romeo on 15/12/9.
//  Copyright © 2015年 heima. All rights reserved.
//

#import "ViewController.h"
#import "HMView.h"

@interface ViewController ()
@property (weak, nonatomic) IBOutlet HMView* hmview;
@property (weak, nonatomic) IBOutlet UISlider* lineWidthProgress;

@end

@implementation ViewController

// 监听线宽改变
- (IBAction)lineWidthChange:(UISlider*)sender
{
    // 把最新的数值(线宽)给了 hmview
    self.hmview.lineWidth = sender.value;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.

    // 默认线宽
    self.hmview.lineWidth = self.lineWidthProgress.value;
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: Interview/Support Files/ViewController.swift
//
//  ViewController.swift
//  Interview
//
//  Created by mac on 2018/1/26.
//  Copyright © 2018年 modi. All rights reserved.
//

import UIKit

class ViewController: UIViewController {
    
    var test: String = ""
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        let p = Person(name: "SaoBing", age: "12")
        p.checkTitleWithAge(age: "12")
        
        UIImageView().image = UIImage(named: "activity")
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
//        test2()
        debugPrint("touchesBegan >>> \(self.classForCoder)")
        super.touchesBegan(touches, with: event)
        
        //测试SEL
        testSelector()
    }
    
    @objc private func test1() {
        var immutableArr = Array<String>()
        debugPrint(unsafeBitCast(immutableArr, to: UnsafeRawPointer.self))
//        debugPrint("im1 = \(immutableArr)")
        immutableArr.append("hello")
        debugPrint(unsafeBitCast(immutableArr, to: UnsafeRawPointer.self))
//        debugPrint("im2 = \(immutableArr)")
        immutableArr.append("hello1")
        debugPrint(unsafeBitCast(immutableArr, to: UnsafeRawPointer.self))
    }
    
    @objc private func test2() {
        let immutableArr = NSMutableArray()
//        debugPrint(unsafeBitCast(immutableArr, to: UnsafeRawPointer.self))
        debugPrint(Unmanaged.passUnretained(immutableArr).toOpaque())
        immutableArr.add("hello")
//        debugPrint(unsafeBitCast(immutableArr, to: UnsafeRawPointer.self))
        debugPrint(Unmanaged.passUnretained(immutableArr).toOpaque())
        immutableArr.add("hello1")
//        debugPrint(unsafeBitCast(immutableArr, to: UnsafeRawPointer.self))
        debugPrint(Unmanaged.passUnretained(immutableArr).toOpaque())
        
    }
    
    private func testSelector() {
        //通过方法名字符串创建一个SEL(Selector)对象
        let sel = Selector.init("test2")
//        perform(sel)
        
        //断言测试
//        assert(sel == nil, "断言生效 ~ ····")
        
        let sel2 = Selector.init(stringLiteral: "test1")
        perform(sel2)
    
        //对比原方法地址与SEL对象地址
        debugPrint("sel ----> \(unsafeBitCast(sel, to: UnsafeRawPointer.self))")
        
        //判断当前类是否实现了某个方法
        let isExist = self.responds(to: "test2")
    }
    
}

